xonsh/xoreutils/tty.py

```diff
@@ -0,0 +1,42 @@
+import os
+import sys
+
+
+def tty(args, stdin, stdout, stderr):
+    if '--help' in args:
+        print(HELP_STR, file=stdout)
+        return 0
+    silent = False
+    for i in ('-s', '--silent', '--quiet'):
+        if i in args:
+            silent = True
+            args.remove(i)
+    if len(args) > 0:
+        if not silent:
+            for i in args:
+                print('tty: Invalid option: {}'.format(i), file=stderr)
+            print("Try 'tty --help' for more information", file=stderr)
+        return 2
+    try:
+        fd = stdin.fileno()
+    except:
+        fd = sys.stdin.fileno()
+    if not os.isatty(fd):
+        if not silent:
+            print('not a tty', file=stdout)
+        return 1
+    if not silent:
+        try:
+            print(os.ttyname(fd), file=stdout)
```

Comment:
As with `cat`, I think this should handle exceptions rather than using guards. As I've often heard, "it's easier to ask forgiveness than to ask permission".

With that in mind, here's the "ERRORS" section of the man page for `ttyname(3p)`:

```
       The ttyname() function may fail if:

       EBADF  The fildes argument is not a valid file descriptor.

       ENOTTY The file associated with the fildes argument is not a terminal.

       The ttyname_r() function may fail if:

       EBADF  The fildes argument is not a valid file descriptor.

       ENOTTY The file associated with the fildes argument is not a terminal.

       ERANGE The  value of namesize is smaller than the length of the string to be returned including
              the terminating null character.
```

Therefore, the standard states that if you get "ENOTTY", it's not a TTY. If you check that, you won't need `os.isatty`.

As with cat too, it's probably worth printing the error's `strerror` if not silent.